WI #616: Addition of five columns related to `Pricing Currency` (#848)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/columns/pricingcurrencycontractedunitprice.md

# Pricing Currency Contracted Unit Price

The Pricing Currency Contracted Unit Price represents the agreed-upon unit price for a single [Pricing Unit](#pricingunit) of the associated SKU, inclusive of [*negotiated discounts*](#glossary:negotiated-discount), if present, while excluding negotiated [*commitment discounts*](#glossary:commitment-discount) or any other discounts. This price is denominated in the [Pricing Currency](#pricingcurrency). The Pricing Currency Contracted Unit Price is commonly used for calculating savings based on negotiation activities. If negotiated discounts are not applicable, the Pricing Currency Contracted Unit Price defaults to the [Pricing Currency List Unit Price](#pricingcurrencylistunitprice).

The PricingCurrencyContractedUnitPrice column adheres to the following requirements:

> * PricingCurrencyContractedUnitPrice MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when the provider supports pricing and billing in different currencies.

Author: Agreement per discussion in Mar 24 Maintainers.

```suggestion
* PricingCurrencyContractedUnitPrice presence in a [*FOCUS dataset*](#glossary:FOCUS-dataset) is defined as follows:
  * PricingCurrencyContractedUnitPrice MUST be present in a FOCUS dataset when the provider supports prices in virtual currency and publishes unit prices exclusive of discounts.
  * PricingCurrencyContractedUnitPrice is RECOMMENDED to be present in a FOCUS dataset when the provider supports pricing and billing in different currencies and publishes unit prices exclusive of discounts.
  * PricingCurrencyContractedUnitPrice MAY be present in a FOCUS dataset in all other cases.
```

> * PricingCurrencyContractedUnitPrice adheres to the following additional requirements:

Author: Thanks for the catch.  Shouldn't be there.
```suggestion
```

* PricingCurrencyContractedUnitPrice MUST be of type Decimal.
* PricingCurrencyContractedUnitPrice MUST conform to [NumericFormat](#numericformat) requirements.
* PricingCurrencyContractedUnitPrice nullability is defined as follows:
  * PricingCurrencyContractedUnitPrice MUST be null when [ChargeCategory](#chargecategory) is "Tax".
  * PricingCurrencyContractedUnitPrice MUST NOT be null when ChargeCategory is "Usage" or "Purchase" and [ChargeClass](#chargeclass) is not "Correction".
  * PricingCurrencyContractedUnitPrice MAY be null in all other cases.
* When PricingCurrencyContractedUnitPrice is not null, PricingCurrencyContractedUnitPrice adheres to the following additional requirements:
  * PricingCurrencyContractedUnitPrice MUST be a non-negative decimal value.
  * PricingCurrencyContractedUnitPrice MUST be denominated in the PricingCurrency.
  * The product of PricingCurrencyContractedUnitPrice, [PricingQuantity](#pricingquantity), and [PricingToBillingExchangeRate](#pricingtobillingexchangerate) MUST match the [ContractedCost](#contractedcost) when PricingQuantity is not null and ChargeClass is not "Correction".

> * Discrepancies in PricingCurrencyContractedUnitPrice, ContractedCost, or PricingQuantity MAY be addressed independently when ChargeClass is "Correction".

Author: This is a copy/paste from the revised normative requirements for `Contracted Unit Price` as authored by @ijurica in PR #798.  Whatever pattern we state there should be repeated here, as the idea is the same.  

Now to your question.  To be honest, I'm not totally sure, as this is ultimately a statement that was drafted in an earlier version.  @ijurica, could you provide some context?

> * Discrepancies in PricingCurrencyContractedUnitPrice, ContractedCost, or PricingQuantity MAY be addressed independently when ChargeClass is "Correction".

Reviewer: This is an example of the original version of that requirement:

> * When ListUnitPrice is present and is not null, multiplying ListUnitPrice by [PricingQuantity](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/columns/listunitprice.md#pricingquantity) MUST equal [ListCost](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/columns/listunitprice.md#listcost), except in cases of ChargeClass "Correction", which may address PricingQuantity or any cost discrepancies independently.

We introduced this "just in case," with the intention to further elaborate on **Correction Handling** later (see [WI 556: Augment support for correction handling](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/issues/556)  ). One thing that comes to mind is addressing Rounding Errors in the PricingQuantity and/or Cost columns, which can also be provided with ChargeCategory set to "Adjustment."

Unfortunately, we weren't able to address WI #556 in this version. As a result, we reworded the requirement in the [PR WI 557: Normative Requirements Refinement - Part 1](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/787), but kept it as-is.

---

I hope that the **Correction Handling augmentation** (including both the attribute and appendix, as well as the refinement of Correction-handling normative requirements in various columns) will be included in the **scope of the next release**.

> * Discrepancies in PricingCurrencyContractedUnitPrice, ContractedCost, or PricingQuantity MAY be addressed independently when ChargeClass is "Correction".

Reviewer 2: I think what we're saying is that, for Corrections, the math might not check out as it would be required by the previous bullet. If that's the case, I think we could reword this for clarity in this changeset. An example alternative wording option, which I would make a sub-bullet of the previous one: 
-  Discrepancies in PricingCurrencyContractedUnitPrice, ContractedCost, or PricingQuantity MAY be present when ChargeClass is "Correction".

> * Discrepancies in PricingCurrencyContractedUnitPrice, ContractedCost, or PricingQuantity MAY be addressed independently when ChargeClass is "Correction".

Author: @gparker-at-sf Please see [this suggestion](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/798/files#r2008011166) on #798 and let me know if that would work -- in which case we would apply that here as well.

# Pricing Currency Contracted Unit Price

> The Pricing Currency Contracted Unit Price represents the agreed-upon unit price for a single [Pricing Unit](#pricingunit) of the associated SKU, inclusive of [*negotiated discounts*](#glossary:negotiated-discount), if present, while excluding negotiated [*commitment discounts*](#glossary:commitment-discount) or any other discounts. This price is denominated in the [Pricing Currency](#pricingcurrency). The Pricing Currency Contracted Unit Price is commonly used for calculating savings based on negotiation activities. If negotiated discounts are not applicable, the Pricing Currency Contracted Unit Price defaults to the [Pricing Currency List Unit Price](#pricingcurrencylistunitprice).

Reviewer 3: ```suggestion
The Pricing Currency Contracted Unit Price represents the agreed-upon unit price for a single [Pricing Unit](#pricingunit) of the associated SKU, inclusive of [*negotiated discounts*](#glossary:negotiated-discount), if present, while excluding negotiated [*commitment discounts*](#glossary:commitment-discount) or any other discounts. This price is denominated in the [Pricing Currency](#pricingcurrency). When negotiated discounts do not apply to unit prices and instead are applied to exchange rates, the Pricing Currency Contracted Unit Price defaults to the [Pricing Currency List Unit Price](#pricingcurrencylistunitprice). The Pricing Currency Contracted Unit Price is commonly used to calculate savings based on negotiation activities. 
```

* PricingCurrencyContractedUnitPrice MUST be of type Decimal.
* PricingCurrencyContractedUnitPrice MUST conform to [NumericFormat](#numericformat) requirements.
* PricingCurrencyContractedUnitPrice nullability is defined as follows:
  * PricingCurrencyContractedUnitPrice MUST be null when [ChargeCategory](#chargecategory) is "Tax".
  * PricingCurrencyContractedUnitPrice MUST NOT be null when ChargeCategory is "Usage" or "Purchase" and [ChargeClass](#chargeclass) is not "Correction".
  * PricingCurrencyContractedUnitPrice MAY be null in all other cases.
* When PricingCurrencyContractedUnitPrice is not null, PricingCurrencyContractedUnitPrice adheres to the following additional requirements:
  * PricingCurrencyContractedUnitPrice MUST be a non-negative decimal value.
  * PricingCurrencyContractedUnitPrice MUST be denominated in the PricingCurrency.
* Discrepancies in PricingCurrencyContractedUnitPrice, ContractedCost, or PricingQuantity MAY exist when ChargeClass is "Correction".

## Column ID

PricingCurrencyContractedUnitPrice

## Display Name

Pricing Currency Contracted Unit Price

## Description

> The agreed-upon unit price for a single Pricing Unit of the associated SKU, inclusive of negotiated discounts, if present, while excluding negotiated commitment discounts or any other discounts, and expressed in Pricing Currency.

Reviewer 3: ```suggestion
The agreed-upon unit price for a single Pricing Unit of the associated SKU, inclusive of *negotiated discounts*, if present, while excluding negotiated *commitment discounts* or any other discounts, and is expressed in Pricing Currency.
```

Some of these issues need to be fixed in the original Contracted Unit Price column as well.

Author: Agree with the addition of asterisks.  Inclusion of "is" is not grammatically correct.  I will revise here, and I have made a note on AI #854 to do a pass on Contracted Unit Price during consistency review (I do not wish to bring that into this PR).

Author: ```suggestion
The agreed-upon unit price for a single Pricing Unit of the associated SKU, inclusive of *negotiated discounts*, if present, while excluding negotiated *commitment discounts* or any other discounts, and expressed in Pricing Currency.
```

  * PricingCurrencyContractedUnitPrice is RECOMMENDED to be present in a FOCUS dataset when the provider supports pricing and billing in different currencies and publishes unit prices exclusive of discounts.
  * PricingCurrencyContractedUnitPrice MAY be present in a FOCUS dataset in all other cases.
* PricingCurrencyContractedUnitPrice MUST be of type Decimal.
* PricingCurrencyContractedUnitPrice MUST conform to [NumericFormat](#numericformat) requirements.
* PricingCurrencyContractedUnitPrice nullability is defined as follows:
  * PricingCurrencyContractedUnitPrice MUST be null when [ChargeCategory](#chargecategory) is "Tax".
  * PricingCurrencyContractedUnitPrice MUST NOT be null when ChargeCategory is "Usage" or "Purchase" and [ChargeClass](#chargeclass) is not "Correction".
  * PricingCurrencyContractedUnitPrice MAY be null in all other cases.
* When PricingCurrencyContractedUnitPrice is not null, PricingCurrencyContractedUnitPrice adheres to the following additional requirements:
  * PricingCurrencyContractedUnitPrice MUST be a non-negative decimal value.
  * PricingCurrencyContractedUnitPrice MUST be denominated in the PricingCurrency.

>   * The product of PricingCurrencyContractedUnitPrice, [PricingQuantity](#pricingquantity), and [PricingToBillingEffectiveExchangeRate](#pricingtobillingeffectiveexchangerate) MUST match the [ContractedCost](#contractedcost) when PricingQuantity is not null and ChargeClass is not "Correction".

Author: Striking this normative requirement, per @ijurica's rationale described [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/848/files#r2018528605).  

```suggestion
```

# Pricing Currency Contracted Unit Price

> The Pricing Currency Contracted Unit Price represents the agreed-upon unit price for a single [Pricing Unit](#pricingunit) of the associated SKU, inclusive of [*negotiated discounts*](#glossary:negotiated-discount), if present, while excluding negotiated [*commitment discounts*](#glossary:commitment-discount) or any other discounts. This price is denominated in the [Pricing Currency](#pricingcurrency). The Pricing Currency Contracted Unit Price is commonly used for calculating savings based on negotiation activities. If negotiated discounts are not applicable, the Pricing Currency Contracted Unit Price defaults to the [Pricing Currency List Unit Price](#pricingcurrencylistunitprice).

Reviewer 4: ```suggestion
The Pricing Currency Contracted Unit Price represents the agreed-upon unit price for a single [Pricing Unit](#pricingunit) of the associated SKU, inclusive of [*negotiated discounts*](#glossary:negotiated-discount), if present, while excluding [*commitment discounts*](#glossary:commitment-discount), negotiated or otherwise, or any other discounts. This price is denominated in the [Pricing Currency](#pricingcurrency). The Pricing Currency Contracted Unit Price is commonly used for calculating savings based on negotiation activities. If negotiated discounts are not applicable, the Pricing Currency Contracted Unit Price defaults to the [Pricing Currency List Unit Price](#pricingcurrencylistunitprice).
```

Author: I lifted this verbiage wholesale from Contracted Unit Price.  So, while I'm not opposed to making the change in a vacuum, I would really prefer the verbiage to stay exactly the same between those two columns.  Therefore, I propose no action here.

Reviewer 4: ok thats fine @shawnalpay

The PricingCurrencyContractedUnitPrice column adheres to the following requirements:

* PricingCurrencyContractedUnitPrice presence in a [*FOCUS dataset*](#glossary:FOCUS-dataset) is defined as follows:
  * PricingCurrencyContractedUnitPrice MUST be present in a FOCUS dataset when the provider presents prices in a virtual currency (e.g. credits, tokens).

>   * PricingCurrencyContractedUnitPrice MUST be present in a FOCUS dataset when the provider presents prices and bills in different national currencies (e.g. priced in USD and billed in EUR).

Reviewer 4: ```suggestion
  * PricingCurrencyContractedUnitPrice MUST be present in a FOCUS dataset when the provider presents prices and usage in different national currencies (e.g. priced in USD and billed in EUR).
```

Author: Same point as I made [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/848#discussion_r1997460587).  Does this work for you, @thecloudman?

```suggestion
  * PricingCurrencyContractedUnitPrice MUST be present in a FOCUS dataset when the provider presents prices and usage in different national currencies (e.g. prices in USD and usage in EUR).
```

  * PricingCurrencyContractedUnitPrice MAY be present in a FOCUS dataset in all other cases.
* PricingCurrencyContractedUnitPrice adheres to the following additional requirements:
* PricingCurrencyContractedUnitPrice MUST be of type Decimal.
* PricingCurrencyContractedUnitPrice MUST conform to [NumericFormat](#numericformat) requirements.
* PricingCurrencyContractedUnitPrice nullability is defined as follows:
  * PricingCurrencyContractedUnitPrice MUST be null when [ChargeCategory](#chargecategory) is "Tax".
  * PricingCurrencyContractedUnitPrice MUST NOT be null when ChargeCategory is "Usage" or "Purchase" and [ChargeClass](#chargeclass) is not "Correction".
  * PricingCurrencyContractedUnitPrice MAY be null in all other cases.
* When PricingCurrencyContractedUnitPrice is not null, PricingCurrencyContractedUnitPrice adheres to the following additional requirements:
  * PricingCurrencyContractedUnitPrice MUST be a non-negative decimal value.

>   * PricingCurrencyContractedUnitPrice MUST be denominated in the BillingCurrency.

Reviewer: ```suggestion
  * PricingCurrencyContractedUnitPrice MUST be denominated in the PricingCurrency.
```

Author: Yes... Thank you @ijurica.  🤦  Rookie mistake.

>   * The product of PricingCurrencyContractedUnitPrice and [PricingQuantity](#pricingquantity) MUST match the [ContractedCost](#contractedcost) when PricingQuantity is not null and ChargeClass is not "Correction".

Reviewer 5: ```suggestion
  * The product of PricingCurrencyContractedUnitPrice, [PricingQuantity](#pricingquantity), and [PricingToBillingExchangeRate](#pricingtobillingexchangerate) MUST match the [ContractedCost](#contractedcost) when PricingQuantity is not null and ChargeClass is not "Correction".
```

>   * The product of PricingCurrencyContractedUnitPrice and [PricingQuantity](#pricingquantity) MUST match the [ContractedCost](#contractedcost) when PricingQuantity is not null and ChargeClass is not "Correction".

Author: I am unsure the precise wording we need for revising this particular line that I lifted from `ContractedUnitPrice`.  I would welcome a suggestion.

Reviewer 5: Wouldn't Contracted Cost = PricingCurrencyContractedUnitPrice * PricingQuantity * PricingToBillingExchangeRate ?

@cnharris10 if one of there is a negotiated discount for a product, I would expect that to be reflected in the PricingCurrencyUnitPrice. For Example, if the a line item is by default 1 token, but the customer has negotiated a discount for this line item to be 0.80 tokens then the PricingCurrencyContractedUnitPrice would be 0.80 tokens.

Author: Resolving via [this committed suggestion](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/848/files/50501c7bac617f215bfcd304eb7f610a42278b9f#r2004389968) from @GrinsteadJ.

> * Discrepancies in PricingCurrencyContractedUnitPrice, ContractedCost, or PricingQuantity MAY be addressed independently when ChargeClass is "Correction".

Reviewer 3: Rather than duplicating all requirements, I wonder if we can just reference ContractedUnitPrice instead and reference that this is in PricingCurrency. Thoughts?

Author: It's _possible_, but there are references to specific columns here, and they're not exactly the same.  

For example, `Contracted Unit Price` has the normative requirement of:

>   * The product of ContractedUnitPrice and [PricingQuantity](#pricingquantity) MUST match the [ContractedCost](#contractedcost) when PricingQuantity is not null and ChargeClass is not "Correction".

And here in `Pricing Currency Contracted Unit Price`, we have:

>   * The product of PricingCurrencyContractedUnitPrice, [PricingQuantity](#pricingquantity), and [PricingToBillingExchangeRate](#pricingtobillingexchangerate) MUST match the [ContractedCost](#contractedcost) when PricingQuantity is not null and ChargeClass is not "Correction".

Those concepts are similar, but the execution is not the same.  And I wouldn't want to try and say, "for these first six requirements in Column Y, refer to Column X, as they are the same -- except for the following requirements, which are different".

While I could get behind the concept of a "clone" column at some point in the future that would let us skip all this duplication, we haven't introduced that precedent before, and I feel uncomfortable with trying to do that on this particular PR.

I will leave this open for a day or two to sustain other thoughts folks have, but I'd prefer to resolve with no action.

## Column ID

PricingCurrencyContractedUnitPrice

## Display Name

Pricing Currency Contracted Unit Price

## Description

The agreed-upon unit price for a single Pricing Unit of the associated SKU, inclusive of negotiated discounts, if present, while excluding negotiated commitment discounts or any other discounts, and expressed in Pricing Currency.

## Usability Constraints

**Aggregation:** Column values should only be viewed in the context of their row and not aggregated to produce a total.

## Content Constraints

| Constraint      | Value                                |
|:----------------|:-------------------------------------|
| Column type     | Metric                               |
| Feature level   | Conditional                          |
| Allows nulls    | True                                 |
| Data type       | Decimal                              |
| Value format    | [Numeric Format](#numericformat)     |
| Number range    | Any valid non-negative decimal value |

## Introduced (version)

> 1.0

Reviewer 3: ```suggestion
1.2
```
